lunatrace/bsl/hasura/migrations/lunatrace/1664218225933_release_analysis_result/up.sql

```diff
@@ -0,0 +1,20 @@
+CREATE TYPE finding_state AS ENUM (
+    'vulnerable',
+    'not_vulnerable'
+    );
+
+CREATE TABLE package.release_analysis_result
+(
+    id                          uuid PRIMARY KEY DEFAULT (gen_random_uuid()),
+    manifest_dependency_edge_id uuid,
+    vulnerability_id            uuid,
+    state                       finding_state,
+    created_at                  timestamptz,
+    source                      TEXT
```

Comment:
what is this "source" column?

Reply:
We know this is a result, but a result of WHAT? What type of analysis was performed? If there are two analysis of the same type, which one was it?  I assume "source" is not an attempt to encapsulate all of that information in a text field. 

I recall from our discussion we decided that analysis types and their definitions would be first class citizens in the database. From that conversation, it was decided that each individual analysis relating to a specific vulnerability (and not just their types) would be represented as a row, to facilitate manual adding/editing of analyses, reviewing of what analyses are possible when viewing a vulnerability, and tracking. 

It follows that for any "auto generated" analyses of the type you envision to populate our early reservoir of analyses, they would be generated at vulnerability insert/upsert time, which recall is a daily job. 

Here is a rough draft of how I think that table would look: 
```sql
CREATE TABLE vulnerability.analysis (
    id                          uuid PRIMARY KEY DEFAULT (gen_random_uuid()),
    auto_generated   bool NOT NULL,
    type                     analysis_type_enum NOT NULL, # excellent application of a table-enum type here, an example row would be "semgrep"
    rule_body jsonb NOT NULL, # having this untyped isnt the end of the world, it needs to be extendable for the different sorts of analysis we want to run. I don't mind a couple more checks to make sure keys arent missing. This can house an entire semgrep rule, no problem. JSONb is truly incredible in PG.
    comment text NULL,
    vulnerability_id uuid NOT NULL,
    package_id uuid NOT NULL,
    semver_range_narrowing text NULL, # we may wish to create rules for package versions that are more specific than what is captured in the vulnerability itself
    skip bool DEFAULT false NOT NULL # not a bad idea to give us a way to short circuit auto generated analyses that might be eroneous without having to hard code it somewhere.
)
    
```

It goes without saying that the time to figure this out is now, not later.

@freeqaz

Reply:
We decided to not implement that yet, coming in a later PR. Agree about having a join table later.

Reply:
Lets meet about that

Reply:
and if you review the document, you can see it's part of v1 here under your items which I think this PR represents
```
Update Database Schema
A place to store the analysis rule for a vulnerability
Change analysis result to point to release_dependency table as the edge
```

Reply:
I think we should discuss this when Chris is back from the dentist. This diff is one step in delivering v0 of the feature, I think this is the minimum to unblock us and is not the endgame schema for the feature.

Reply:
> what is this "source" column?

Source would be the name of the scanner that generated the result. Open to changing to enum but I left it as text for now.